feat: [Subscales] Add Severity to Lookup Table (CAMHI) (M2-7586)

sultanofcardio opened the pull request:
- [x] Tests for the changes have been added

### 📝 Description

🔗 [Jira Ticket M2-7586](https://mindlogger.atlassian.net/browse/M2-7586)

> [!NOTE]
> This feature is behind the flag `enableCahmiSubscaleScoring`

This PR adds support for the severity column in the subscales lookup table. This includes the following changes:
- The subscales lookup table template now includes a severity column
- The parsing code has been updated to support the proper severity values (Minimal, Mild, Moderate, Severe) and the empty string
- The age column in the lookup table now supports ranges using the interval symbol ~ (e.g. 10~15). This is similar to the ranges supported by the `score` and `rawScore` columns
- The lookup table preview has been updated to show the severity column
- The severity values are being saved to the database


### 📸 Screenshots

#### CSV Template

<img width="360" alt="image" src="https://github.com/user-attachments/assets/28839470-9a42-4fdf-a53a-338b90442b47"> <img width="360" alt="image" src="https://github.com/user-attachments/assets/53dea10a-6525-4c82-a02a-015de3f969ef">

#### Lookup Table Preview

<img width="1552" alt="image" src="https://github.com/user-attachments/assets/ba65e869-1f8e-44ed-bec5-9b09b2e8ca36">

<img width="1552" alt="image" src="https://github.com/user-attachments/assets/6984d9bf-8b66-4b72-972c-2ae087b4824c">

### 🪤 Peer Testing

1. Edit an existing activity
2. Add an item with a score attached (such as a Single Selection item)
3. Go to the Activity Settings tab
4. Click Subscales Configuration
5. Create a subscale
6. Click the little <img width="30" alt="image" src="https://github.com/user-attachments/assets/3bd85f65-c17c-4075-84ec-1f1af4e17a92"> icon
7. Download the template
8. Re-upload the template without editing
9. Confirm that the preview includes the severity column, and a warning is shown for the incomplete severity column
10. Click save on the modal
11. Click the Save & Publish button
12. Check the database and confirm the presence of the `severity` property in the `subscale_setting` JSONB column of the `activities` table
    <img width="332" alt="image" src="https://github.com/user-attachments/assets/d580b886-1987-4896-98fe-256200e9a7e0">
13. Repeat the test with invalid severity values and confirm that parsing fails
14. Turn off the feature flag and confirm that the current behaviour functions as expected

### ✏️ Notes

1. ~Helper text that explains the various severity levels is coming in the follow-up ticket M2&#8722;7747~

sultanofcardio committed:
Add subscales feature flag

sultanofcardio committed:
Add a regex generator function

sultanofcardio committed:
Update parsing logic to support severity column

sultanofcardio committed:
Update csv template file

sultanofcardio committed:
Incorporate feature flag

sultanofcardio committed:
Add severity data to table preview

sultanofcardio committed:
Add incomplete severity data warning

sultanofcardio committed:
Fix feature flag passing

sultanofcardio committed:
Fix broken tests

sultanofcardio committed:
Merge branch 'refs/heads/develop' into feature/M2-7586-lookup-table-severity

farmerpaul commented on src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/LookupTable/LookupTable.hooks.ts:
```diff
@@ -29,7 +40,29 @@ export const useSubscaleLookupTableSetup = ({
+    const incompleteSeverityData = mappedData.some((item) => !item.severity);
+    if (incompleteSeverityData) {
+      setWarning(warnings.incompleteSeverityData);
+    }
```
Should these lines also be behind the feature flag? I'm getting this when it's turned off, and I upload a CSV that lacks the `severity` column:
![CleanShot 2024-09-16 at 17 32 15@2x](https://github.com/user-attachments/assets/12709ab1-51af-46bb-8e7c-fb778f3d5816)

sultanofcardio replied:
Thanks for catching this Paul 🙏

farmerpaul reviewed (approved):
Just discovered a minor bug with the feature flag turned off, but seems like a simple fix, so preapproving. All other AC behave as expected. :+1:

sultanofcardio committed:
Support tilde ranges in the subscale lookup file

mbanting commented on src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/LookupTable/LookupTable.hooks.ts:
Is this where tests will be added for complete and incomplete severity data?

sultanofcardio replied:
Thanks for reviewing Marty. I've been adding my tests to this file https://github.com/ChildMindInstitute/mindlogger-admin/blob/e45814f1f8bd90f9f34d5df75403f138221ae56a/src/modules/Builder/features/ActivitySettings/SubscalesConfiguration/SubscalesConfiguration.test.tsx#L185

sultanofcardio replied:
@Marty tests have now been added

mbanting replied:
Thanks!

mbanting reviewed (commented):
Great job @sultanofcardio . 
Did some quick testing and it works well. Just the feature flag on the severity check, as @farmerpaul pointed out. 

Noticed in the description that tests are still in progress. I can have a look again once those are added but PR is looking great so far.

sultanofcardio committed:
Wrap incomplete severity data warning in `enableCahmiSubscaleScoring` feature flag

sultanofcardio committed:
Update age regex to support 0

sultanofcardio committed:
Feature flag age range support

sultanofcardio committed:
Add some tests

sultanofcardio committed:
Move age range regex to utils file

sultanofcardio commented:
@mbanting and @farmerpaul I've added some tests, and updated the code to address the new AC for supporting age ranges. Please give it a re-review when you have a minute 🙏

mbanting reviewed (approved):
Looks great @sultanofcardio . Good set of tests. 
Noticed that it's possible for the age range to be 10~5 (ie. first number large than second number). It wasn't called out in the AC so happy to approve here but raised it in a [comment](https://mindlogger.atlassian.net/browse/M2-7586?focusedCommentId=24911) on the ticket, suggesting we tackle this in a followup.

sultanofcardio committed:
Fix broken argument passing
